lib/elixir/lib/module/types/pattern.ex

```diff
@@ -261,7 +264,8 @@ defmodule Module.Types.Pattern do
     # should not affect the inference of x
     if not type_guard? or consider_type_guards? do
       stack = push_expr_stack(expr, stack)
-      param_unions = signature_to_param_unions(signature, context)
+      expected_signature = filter_clauses(signature, expected, stack, context)
```

Comment:
We shouldn't filter clauses twice. :) We should pass the filtered signature to unify_args.